docs/tutorials/tutorial_2.rst

```diff
@@ -0,0 +1,548 @@
+Tutorial 2: Minimum Spanning Tree
+=================================
+
+In this tutorial, you will learn how to write a generator for a more complicated problem, called **Minimum Spanning Tree** with test groups. We will focus on the general idea of using test groups in **tcframe**.
+
+Here is the complete problem statement:
+
+----
+
+| **Minimum Spanning Tree**
+| Time Limit: 1 second
+| Memory Limit: 64 MB
+|
+| **Description**
+| Given a weighted, connected graph G consisting of N nodes and M edges. The nodes are numbered from 0 to N - 1. You need to choose N - 1 out of those M edges such that the graph remains connected if we only consider the chosen edges. What is the minimum possible sum of the weight of the chosen edges?
+|
+| **Input Format**
+| The first line contains two integers N and M separated by a space.
+| M lines follow. Each line contains three integers U, V, and W separated by a space, representing an edge connecting node U and V with weight W.
+|
+| **Output Format**
+| A single line containing the minimum possible sum of the weight of the chosen edges.
+|
+| **Sample Input**
+| ``3 3``
+| ``0 1 2``
+| ``1 2 3``
+| ``2 0 4``
+|
+| **Sample Output**
+| ``5``
+|
+| **Constraints for all subtasks**
+| 2 ≤ N ≤ 100,000
+| N - 1 ≤ M ≤ 100,000
+| 0 ≤ U, V < N
+| 1 ≤ W ≤ 1,000
+| The graph is connected
+| There is no edge connecting a node to itself
+|
+| **Subtask 1**
+| M ≤ 20
+|
+| **Subtask 2**
+| All W are equal
+|
+| **Subtask 3**
+| No additional constraints
+|
+
+----
+
+Since the steps of preparing the problem package and writing the solution are similar to the previous tutorial, we will not cover those steps in this tutorial. Instead, we will focus on writing the problem spec and test spec. Let's use ``minimum-spanning-tree`` as the slug for this problem.
+
+----
+
+Writing problem spec
+--------------------
+
+The following is a problem spec that is derived directly from the problem statement. We are using the same private method ``eachElementBetween()`` from the previous tutorial. We are adding another private method ``allAreEqual()``. We are also adding several private methods for checking graph properties, namely ``noSelfLoop()`` (self loop is an edge connecting a node to itself) and ``isConnected()``. These methods are very modular and we can always reuse them for another problem. However, note that the methods assume the nodes of the graph are 0-based, and several minor changes are needed if the nodes are 1-based.
+
+Let's focus on the syntax differences compared to the first tutorial. Note that we are using the method ``Constraints()`` to include the global constraints that applies to every subtasks, and we are using ``SubtaskX()`` to include the constraints that applies to only those subtasks. Since ``Subtask3()`` does not have any additional constraints that is not included in the global constraints, we can simply just leave the method empty.
```

Comment:
- that applies -> that apply
- every subtasks -> every subtask
- is not included -> are not included